airflow/utils/metastore_cleanup.py

```diff
@@ -0,0 +1,289 @@
+    entries_to_delete = query.limit(max_rows_to_print).all()
+    logging.debug("print entities query: " + str(query))
+    for entry in entries_to_delete:  # type: Log
+        print(entry.__dict__)
+
+
+def do_delete(query):
+    print("Performing Delete...")
+    # using bulk delete
+    query.delete(synchronize_session=False)
+    session.commit()
+    print("Finished Performing Delete")
+
+
+def subquery_keep_last(keep_last_filters, keep_last_group_by):
+    # workaround for MySQL "table specified twice" issue
+    # https://github.com/teamclairvoyant/airflow-maintenance-dags/issues/41
+    subquery = session.query(func.max(DagRun.execution_date))
+
+    if keep_last_filters is not None:
+        for entry in keep_last_filters:
+            subquery = subquery.filter(entry)
+
+    if keep_last_group_by is not None:
+        subquery = subquery.group_by(keep_last_group_by)
+
+    subquery = subquery.from_self()
+    return subquery
+
+
+def build_query(
+    orm_model, recency_column, keep_last, keep_last_filters, keep_last_group_by, clean_before_timestamp
+):
+    query = session.query(orm_model)
+    conditions = [recency_column < clean_before_timestamp]
+    if keep_last:
+        subquery = subquery_keep_last(
+            keep_last_filters=keep_last_filters, keep_last_group_by=keep_last_group_by
+        )
+        conditions.append(recency_column.notin_(subquery))
+    query = query.filter(and_(*conditions))
+    return query
+
+
+logger = logging.getLogger(__file__)
+
+
+def _cleanup_table(
+    orm_model,
+    recency_column,
+    keep_last,
+    keep_last_filters,
+    keep_last_group_by,
+    clean_before_timestamp,
+    dry_run=True,
+    verbose=False,
+):
+    print()
+    if dry_run:
+        print(f"Performing dry run for table {orm_model.__tablename__!r}")
+    query = build_query(
+        orm_model=orm_model,
+        recency_column=recency_column,
+        keep_last=keep_last,
+        keep_last_filters=keep_last_filters,
+        keep_last_group_by=keep_last_group_by,
+        clean_before_timestamp=clean_before_timestamp,
+    )
+
+    print_entities(query, print_rows=False)
+
+    if not dry_run:
+        do_delete(query)
+
+
+def confirm(date, tables):
+    for_tables = f" for tables {tables!r}" if tables else ''
+    question = '\n'.join(
+        [
+            f"You have requested that we purge all data prior to {date}{for_tables}.",
+            "This is irreversible.  Consider backing up the tables first and / or doing a dry run with option --dry-run.",
+            "Enter 'delete rows' to proceed.",
+        ]
+    )
+    print(question)
+    answer = input().strip()
+    if not answer == 'delete rows':
+        SystemExit("User did not confirm; exiting.")
+
+
+def run_cleanup(clean_before_timestamp: DateTime, table_names=None, dry_run=False, verbose=False):
+    clean_before_timestamp = timezone.coerce_datetime(clean_before_timestamp)
+    if not dry_run:
+        confirm(clean_before_timestamp, table_names)
```

Comment:
Thoughts on adding a way to bypass this?

 For example, if we wanted to have a DAG or CronJob which ran a cleanup command daily? I know lots of other CLIs have a `--yes` flag to skip the confirmation step.

Reply:
yup it's a good idea i already added a `confirm` flag so that we can use from python (without user input ) but skipping from cli also  good option thanks